### ISSUE TITLE
allParams() is empty when sending data from an IOS app

### ISSUE THREAD

**Issue by samueljoos** [balderdashy/skipper#50]
Hi,

When I try to send multipart data with post data from an IOS app to sailsjs the the params are alway empty. I use this Swift code for sending the data:

```
        let manager:AFHTTPRequestOperationManager = AFHTTPRequestOperationManager()
        let parameters:NSDictionary = [
            "title":"test title",
            "body":"test body"
        ]
        
        let imageData = UIImageJPEGRepresentation(UIImage(named: "image"), 0.5)
        
        manager.POST("http://domain.com/post/create",
            parameters: parameters,
            constructingBodyWithBlock: { (formData) -> Void in
                let afFormData:AFMultipartFormData = formData
                afFormData.appendPartWithFileData(imageData, name: "image", fileName: "image.jpg", mimeType: "image/jpeg")
            },
            success: { (request, result) -> Void in
                Swell.info(result)
            }) { (request, error) -> Void in
                Swell.info(error)
        }
```

on the sailsjs side I use this:
```
  create: function (req, res) {
        var params = req.allParams();
        return res.json(params);
  }
```

But the params are received when using the same data in a simple html form.

When I monitor the request with Charles Proxy I get this:
https://www.dropbox.com/s/jp4i3vz1ka6vfli/Screenshot%202014-11-12%2010.22.36.png?dl=0
So the data is definitely being send.

**Comment by mikermcneil** [comment]
@mrtavo @samueljoos maybe the default swift implementation doesn't send file parameters last?  To send files and text body params in the same request with skipper, you'll want to make sure the text params come _first_, before the body param.  Hope that helps! (also please try out skipper@0.5.4 and skipper-disk@0.5.4 and let me know if things get better)

**Comment by blackbumer** [comment]
I use skipper-gridfs and have same problem. I tried use postman with Text parameters first and it didn't help

**Comment by blackbumer** [comment]
It works on local host but doesn't work on remote host

**Comment by blackbumer** [comment]
I made a test app on Heroku
here is controller code
    post: function(req, res, next) {
        debug("Upload asset");
        var md5 = req.body.md5;
        var aType = parseInt(req.body.aType);
        debug("md5 [%s], asset type (%s)", md5, aType);
        res.json({
            md5: md5,
            aType: aType
        });
    },

You can use postman to check it. Once you add file at the end of the form this method stops to detect params in body

https://agile-refuge-3063.herokuapp.com/asset

**Comment by blackbumer** [comment]
```
2015-02-23T05:44:03.092256+00:00 heroku[web.1]: State changed from starting to up
2015-02-23T05:44:16.478054+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router dispatching POST /asset
2015-02-23T05:44:16.480690+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT body-parser:json content-type "multipart/form-data; boundary=----WebKitFormBoundaryRttYXpKRm0mePTwm"
2015-02-23T05:44:16.482930+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT body-parser:urlencoded skip parsing
2015-02-23T05:44:16.483007+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router _parseHTTPBody  : /asset
2015-02-23T05:44:16.479891+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router expressInit  : /asset
2015-02-23T05:44:16.539251+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT skipper passed control to app because 50ms elapsed
2015-02-23T05:44:16.547085+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT skipper undefined
2015-02-23T05:44:16.549734+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router serveStatic  : /asset
2015-02-23T05:44:16.479393+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router query  : /asset
2015-02-23T05:44:16.480334+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router jsonParser  : /asset
2015-02-23T05:44:16.482181+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT body-parser:json skip parsing
2015-02-23T05:44:16.482311+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router urlencodedParser  : /asset
2015-02-23T05:44:16.482537+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT body-parser:urlencoded content-type "multipart/form-data; boundary=----WebKitFormBoundaryRttYXpKRm0mePTwm"
2015-02-23T05:44:16.539483+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT skipper waiting for any text params
2015-02-23T05:44:16.549217+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router <anonymous>  : /asset
2015-02-23T05:44:16.550208+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router trim prefix (/asset) from url /asset
2015-02-23T05:44:16.550440+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router dispatching POST /
2015-02-23T05:44:16.551285+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT rest Post asset
2015-02-23T05:44:16.550301+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT express:router router /asset : /asset
2015-02-23T05:44:16.551410+00:00 app[web.1]: Mon, 23 Feb 2015 05:44:16 GMT rest md5 [undefined], asset type (NaN)
2015-02-23T05:44:16.571895+00:00 heroku[router]: at=info method=POST path="/asset" host=agile-refuge-3063.herokuapp.com request_id=b8853c98-e22e-4141-9093-39257452a08c fwd="174.58.226.76" dyno=web.1 connect=1ms service=95ms status=200 bytes=179

**Comment by blackbumer** [comment]
this fix helped me 
https://github.com/blackbumer/skipper/commit/364a4153a6fcd1f053c70f55e3d735868035a55d

**Comment by sailsbot** [comment]
Thanks for posting, @samueljoos.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
